OSDOCS-18910: Document OLM v1 deployment configuration API (#111495)
openshift/openshift-docs: modules/olmv1-deployment-config-api.adoc

// Module included in the following assemblies:
//
// * extensions/ce/olmv1-configuring-extensions.adoc

:_mod-docs-content-type: CONCEPT
[id="olmv1-deployment-config-api_{context}"]
= `deploymentConfig` API

[role="_abstract"]
The deploymentConfig API provides runtime configuration for operator pods, offering feature parity with {olmv0} Subscription configuration.

:FeatureName: {olmv1} `deploymentConfig` API
include::snippets/technology-preview.adoc[]

Provides feature parity with {olmv0}'s `Subscription.spec.config`. Configure resources, node placement, storage, environment variables, and other deployment settings.

[id="olmv1-deployment-config-structure_{context}"]
== `deploymentConfig` structure

Configure how the operator deploys in the `spec.config.inline.deploymentConfig` field as a JSON object.

.Example `deploymentConfig` object
[source,yaml]
----
apiVersion: olm.operatorframework.io/v1
kind: ClusterExtension
metadata:
  name: <extension_name>

> spec:

Reviewer: **Missing required fields in YAML example**

This example omits `serviceAccount` and `source`, both of which are required fields in `ClusterExtensionSpec`. While this is labeled as showing just the `deploymentConfig` structure, users may copy-paste it and hit validation errors. Consider adding them (even as placeholders like `<service_account_name>` and a minimal `source` block), or adding a note that these fields are required but omitted for brevity.

  namespace: <installation_namespace>
  config:
    configType: Inline
    inline:
      deploymentConfig:
        resources:
          requests:
            cpu: 100m
            memory: 128Mi
          limits:
            cpu: 500m
            memory: 512Mi
        nodeSelector:
          node-role.kubernetes.io/infra: ""
        tolerations:
        - key: node-role.kubernetes.io/infra
          operator: Exists
          effect: NoSchedule
----
+
where:
+
--
`deploymentConfig:`:: Object for customizing the deployment.
`resources:`:: CPU and memory requests and limits.
`nodeSelector:`:: Node placement selector.
`tolerations:`:: Node taint tolerations.
--

[id="olmv1-deployment-config-fields_{context}"]
== Supported configuration fields

> Environment variables:: Add or override environment variables with `env` and `envFrom`. Values are merged with existing container environment variables, with `deploymentConfig` values taking precedence.

Reviewer: **Inaccuracy: `env` and `envFrom` have different merge behaviors**

This line groups `env` and `envFrom` together and says both are "merged with `deploymentConfig` values taking precedence," but that's only true for `env`.

Per the code:
- **`env`** ([`generators.go:619-644`](https://github.com/openshift/operator-controller/blob/main/internal/operator-controller/rukpak/render/registryv1/generators/generators.go#L619-L644)): overrides existing env vars by name, appends new ones. Config takes precedence. ✅ matches
- **`envFrom`** ([`generators.go:650-672`](https://github.com/openshift/operator-controller/blob/main/internal/operator-controller/rukpak/render/registryv1/generators/generators.go#L650-L672)): **appended** with deduplication. No precedence/override behavior.

Suggested fix: describe them separately, e.g.:
```
Environment variables:: Add or override environment variables with `env`. Values are merged with existing container environment variables, with `deploymentConfig` values taking precedence.

Environment variable sources:: Add environment variable sources with `envFrom`. Sources are appended to existing sources; duplicates are skipped.
```

Resource requirements:: Specify CPU and memory requests and limits with `resources`. Replaces existing resource requirements.

Node selector:: Control pod node placement with `nodeSelector`. Replaces existing node selector.

Tolerations:: Schedule pods on nodes with taints by using `tolerations`. Appended to existing tolerations.

Affinity rules:: Define pod affinity and anti-affinity rules with `affinity`. Non-nil fields replace corresponding bundle fields.

> Volumes and volume mounts:: Add `emptyDir`, `configMap`, or `secret` volumes. Appended to existing volumes.

Reviewer: **Inaccuracy: volume type restriction and merge behavior**

Two issues here:

1. **No volume type restriction exists.** The `volumes` field accepts `[]corev1.Volume` — any Kubernetes volume type is valid. The schema generator (`hack/tools/schema-generator/main.go`) does not filter volume types. Claiming only `emptyDir`, `configMap`, or `secret` are supported will mislead users into thinking other types are rejected.

2. **Behavior is not "appended".** Per [`generators.go:678-695`](https://github.com/openshift/operator-controller/blob/main/internal/operator-controller/rukpak/render/registryv1/generators/generators.go#L678-L695), volumes **override** existing entries with the same name, and only append if no volume with that name exists. The code comment explicitly notes this differs from OLMv0:
   ```go
   // This differs from OLMv0, which appends volumes without checking for duplicates
   ```

Suggested fix:
```
Volumes and volume mounts:: Add volumes and volume mounts. Volumes with the same name as existing bundle volumes are overridden; new volumes are appended.
```

> Annotations:: Add custom pod annotations. Merged with existing annotations, with bundle values taking precedence on conflicts.

Reviewer: **Suggestion: clarify what "bundle values taking precedence" means in practice**

The wording is technically correct — per [`generators.go:880-904`](https://github.com/openshift/operator-controller/blob/main/internal/operator-controller/rukpak/render/registryv1/generators/generators.go#L880-L904), existing annotations are preserved and config annotations only fill in keys that don't already exist. But "bundle values taking precedence on conflicts" could be read as "your annotations are merged and conflicts are resolved" when the real implication is: **you cannot override annotations that the bundle already sets.**

Suggested clarification:
```
Annotations:: Add custom pod annotations. Merged with existing annotations. Annotations already set by the bundle cannot be overridden.
```

// Module included in the following assemblies:
//
// * extensions/ce/olmv1-configuring-extensions.adoc

:_mod-docs-content-type: CONCEPT
[id="olmv1-deployment-config-api_{context}"]

> = `deploymentConfig` API

Reviewer 2: ```suggestion
= deploymentConfig API
```

[role="_abstract"]
The `deploymentConfig` API controls Operator pod runtime settings such as resources, node placement, and environment variables, providing feature parity with {olmv0} Subscription configuration.

:FeatureName: {olmv1} `deploymentConfig` API
include::snippets/technology-preview.adoc[]

> Provides feature parity with {olmv0}'s `Subscription.spec.config`. Configure resources, node placement, storage, environment variables, and other deployment settings.

Reviewer 2: ```suggestion
Provides feature parity with the `Subscription.spec.config` configuration of the {olmv0}. Configure resources, node placement, storage, environment variables, and other deployment settings.
```

<img width="643" height="246" alt="Image" src="https://github.com/user-attachments/assets/ee212db2-7672-4f8c-bc2d-ecaab0ba0e95" />

[id="olmv1-deployment-config-structure_{context}"]

> == `deploymentConfig` structure

Reviewer 2: Same backtick in heading question.

Affinity rules:: Define pod affinity and anti-affinity rules with `affinity`. Non-nil fields replace corresponding bundle fields.

Volumes and volume mounts:: Add volumes and volume mounts. Volumes with the same name as existing bundle volumes are overridden; new volumes are appended.

Annotations:: Add custom pod annotations. Merged with existing annotations. Annotations already set by the bundle cannot be overridden.

[id="olmv1-deployment-config-validation_{context}"]
== Configuration validation

{olmv1} validates configuration against a JSON schema generated from Kubernetes API definitions. The schema derives from the `SubscriptionConfig` type used in {olmv0}, providing consistent validation across versions.

Invalid configurations prevent installation and report errors in the `ClusterExtension` resource's `Progressing` condition. Common validation errors include:

* Unknown field errors when using unsupported configuration options
* Type mismatch errors when field values do not match the expected type
* Required field errors when mandatory nested fields are missing

[NOTE]
====

> {olmv1} applies configurations during the ClusterObjectSet deployment process, modifying Operator manifests before organizing them into phases.

Reviewer 2: ```suggestion
{olmv1} applies configurations during the `ClusterObjectSet` deployment process, modifying Operator manifests before organizing them into phases.
```
